### PR TITLE
feat(babel): add babel-preset-macros

### PR DESCRIPTION
**Pull request by kentcdodds**
Because 😎 it's awesome.

giphy

### CODE REVIEW

File: build/babel/preset.js [hunk 1]
```diff
@@ -44,6 +44,7 @@ module.exports = (context, opts = {}) => ({
     }]
   ],
   plugins: [
+    require('babel-preset-macros'),
```
Comment (tricoder42) [inline build/babel/preset.js]: Small typo here: It's not a preset, but plugin ;)

Comment (tricoder42) [reply build/babel/preset.js]: NOTE: I'm not stalking you, I was just curious this morning, why macros aren't included in Next.js 7 😄

Comment (kentcdodds) [reply build/babel/preset.js]: lol Whoops!

**Comment by HaNdTriX** [comment]
It is indeed! ♥️

Since next.js already has an example for them, we might want to update that as well.

* [next.js/examples/with-babel-macros](https://github.com/zeit/next.js/tree/canary/examples/with-babel-macros)

**Comment by kentcdodds** [comment]
Updating the example now.

**Comment by kentcdodds** [comment]
Alrighty, this PR has been updated and the example is pretty sweet I think.

File: examples/using-babel-plugin-macros/pages/posts-custom/import-posts.macro.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
+// this is a more fine-tuned version of import-all.macro for our specific use-case.
```
Comment (lucleray) [inline examples/using-babel-plugin-macros/pages/posts-custom/import-posts.macro.js]: Files in the `pages/` directory are exported as pages, so we might want to move this to a different directory (`lib/` maybe ?)

Comment (kentcdodds) [reply examples/using-babel-plugin-macros/pages/posts-custom/import-posts.macro.js]: Ah, fair point!

Comment (kentcdodds) [reply examples/using-babel-plugin-macros/pages/posts-custom/import-posts.macro.js]: Actually no, I want `posts-custom` to be a page. This is good as-is 👍

Comment (lucleray) [reply examples/using-babel-plugin-macros/pages/posts-custom/import-posts.macro.js]: This file will generate `posts-custom/import-posts.macro` page, not `posts-custom` 😅

Comment (lucleray) [reply examples/using-babel-plugin-macros/pages/posts-custom/import-posts.macro.js]: Yep, every files under pages are considered a page

File: examples/using-babel-plugin-macros/pages/posts-custom/index.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import Link from 'next/link'
+
+// the import-all.macro that we're using in /posts is cool, but the drawback is
+// that we're literally including all of the data/code for all the posts in that
+// file which may not be what we want.
+//
+// So let's write our own macro that's specific for our use case which will
+// replace all instances of `allPosts` with only data we need for this page
+// (path, and title).
+import allPosts from '../lib/import-posts.macro'
```
Comment (lucleray) [inline examples/using-babel-plugin-macros/pages/posts-custom/index.js]: The relative path is not correct here , it should be `../../lib/import-posts.macro`

**Review by timneutkens — COMMENTED** [review]
While macros are cool and useful in some cases, I don't think we should add them into the Next.js core, it's already possible to add the macros plugin to `.babelrc` and have them work. We're also working on reducing the number of core dependencies that Next.js ships with.

**Pull request closed without merging**